core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala

```diff
@@ -139,6 +199,356 @@ protected[actions] trait PrimitiveActions {
+                        case Failure(t) =>
+                          // resolution failure
+                          val response =
+                            ActivationResponse.applicationError(compositionComponentNotFound(next.toString))
+                          completeAppActivation(user, session, response)
+                        case Success(next) =>
+                          if (session.accounting.components >= actionSequenceLimit) {
+                            // composition is too long
+                            val response = ActivationResponse.applicationError(compositionIsTooLong)
+                            completeAppActivation(user, session, response)
+                          } else {
+                            // successful resolution
+                            invokeComponent(user, next, params, session)
+                          }
+                      }
+                  }
+              }
+          }
+      }
+    }
+  }
+
+  /**
+   * A method that knows how to handle a component invocation.
+   *
+   * This method distinguishes primitive actions, sequences, and compositions.
+   * If the component action is a composition, it invokes invokeComposition and returns.
+   * invokeComposition takes care of resuming the execution of this composition, when the nested composition finishes.
+   * If the component action is not a composition, it is invoked followed by the reinvocation of the conductor action.
+   * This method also keeps track of the duration and memory footprint for the composition.
+   *
+   * @param user the identity invoking the action
+   * @param action the component action to invoke
+   * @param payload the dynamic arguments for the activation
+   * @param session the session object for this composition
+   * @param transid a transaction id for logging
+   */
+  private def invokeComponent(user: Identity, action: WhiskActionMetaData, payload: Option[JsObject], session: Session)(
+    implicit transid: TransactionId) {
+
+    val exec = action.toExecutableWhiskAction
+    if (action.annotations.get("conductor").isDefined && exec.isDefined) {
+      // composition
+      invokeComposition(user, exec.head, payload, None, Some(session.activationId), Some(session)) // non-blocking
+      // invokeComposition will take care of accounting and continuations
+    } else {
+      session.accounting.components += 1
+      exec
+        .map { exec =>
+          // not a sequence, not a composition
+          invokeSimpleAction(
+            user,
+            exec,
+            payload,
+            Some(action.limits.timeout.duration + 1.minute),
+            Some(session.activationId))
+        }
+        .getOrElse {
+          // sequence
+          val SequenceExecMetaData(components) = action.exec
+          invokeSequence(user, action, components, payload, None, Some(session.activationId), false, 0).map(r => r._1)
+        }
+        .onComplete {
+          case Failure(t) =>
```

Comment:
this blobs seems repetitive with a previous on complete - can this be consolidated?

Reply:
I refactored the code.